### PR TITLE
Search backend: expand glob refs as part of repo resolution ref validation

### PR DESCRIPTION
**Pull request by camdencheek**
This adds an "expand ref globs" step to repo resolution. This is something that we always end up doing anyways, but it's spread across the codebase closer to gitserver accesses. In a quite a few places, we don't actually handle globs, we just punt or assume that there are no globs at that step, which is somtimes not true. This change encodes the expansion into the type system by making globs not representable in the `RepositoryRevisions` type.

This change is important because until we've expanded glob patterns, we _don't know what revision we're working with_ because a glob usually expands to multiple revisions or no revisions. As an example of this, right now, `repo:has.commit.after(date)` does not work with globs because the presence of a ref glob means don't know what revision to actually do the check for. For the same reason, this is needed for the optimization of general `repo:contains...` optimizations. We need to know which refs we're filtering from in order to filter correctly.

This simplifies the `RepositoryRevisions` struct so that it no longer can contain globs. As part of this, I got to get rid of all the lazy glob expansion that is filling that struct and make it just a pure data type.

Again, sorry for the large change, but I was unable to figure out how to do this any more incrementally. 

## Test plan

All existing tests pass. Did some manual testing to feel better about it.

### CODE REVIEW

File: internal/search/alert.go [hunk 1]
```diff
@@ -157,74 +157,6 @@ func AlertForStructuralSearchNotSet(queryString string) *Alert {
 	}
 }
 
-func AlertForMissingRepoRevs(missingRepoRevs []*RepositoryRevisions) *Alert {
```
Comment (camdencheek) [inline internal/search/alert.go:160]: Moved this to where it is used (`internal/search/alert`) to fix circular dependency

File: internal/search/repo_revs.go [hunk 2]
```diff
@@ -59,32 +54,15 @@ func (r1 RevisionSpecifier) Less(r2 RevisionSpecifier) bool {
...
-
-	// resolveErr stores the error returned by the first call to ExpandedRevSpecs. It
-	// gives the caller the chance to distinguish between an error and an empty resolvedRevs.
-	resolveErr error
-
-	// ListRefs is called to list all Git refs for a repository. It is intended to be mocked by
-	// tests. If nil, gitserver.ListRefs is used.
-	ListRefs func(ctx context.Context, repo api.RepoName) ([]gitdomain.Ref, error) `json:"-"`
+	Revs []string
 }
```
Comment (camdencheek) [inline internal/search/repo_revs.go:58]: Primary change: `RepositoryRevisions` is now a data-only type that contains a repository and a set of revisions. These revisions cannot be globs at this point. The `RevisionSpecifier` type still exists to represent the input values values for `rev`, but to create this type, globs must be expanded into revisions.

File: internal/search/repos/repos.go [hunk 3]
```diff
@@ -922,3 +989,21 @@ func PrivateReposForActor(ctx context.Context, logger log.Logger, db database.DB
 	}
 	return userPrivateRepos
 }
+
+type RepoRevSpecs struct {
+	Repo types.MinimalRepo
+	Revs []search.RevisionSpecifier
+}
```
Comment (camdencheek) [inline internal/search/repos/repos.go:996]: This type replaces the purpose of the previous `search.RepositoryRevisions` type. It still exists to represent the state of "parsed, but not expanded" which only really exists in this file. The only reason it is exported is so we can use it for missing revisions alerts.

**Comment by keegancsmith** [comment]
FYI an important perf optimization we do when just hitting zoekt is if zoekt has the revs we don't speak to gitserver. Haven't looked at code, but just want to make sure we don't regress on that.

**Comment by keegancsmith** [comment]
Historically another reason we did this lazily was because we could start off searcher as soon as we resolved the ref for a repo. Rather than blocking until all refs are resolved for all repos.

**Comment by camdencheek** [comment]
@keegancsmith thanks for hopping in. It's certainly possible I'm misunderstanding this, but I think the following makes sense. 

> if zoekt has the revs we don't speak to gitserver

I don't think this is true right now unless we're doing a global search, and we don't resolve repos at all for that. For each resolved repo+revision, [unless the revision is a glob or HEAD](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@cb906941e8b49806f0be1861fe025b9de863ecdb/-/blob/internal/search/repos/repos.go?L342-352), we [check that the revision exists](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@cb906941e8b49806f0be1861fe025b9de863ecdb/-/blob/internal/search/repos/repos.go?L369-369). 

Additionally, how would it even be possible to know if zoekt has the rev if a user gave a glob pattern? Zoekt cannot resolve globs, so it can't report whether it matched. In fact, right now, it looks like if a repo has any glob revs, we [mark the whole thing as unindexed](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@cb906941e8b49806f0be1861fe025b9de863ecdb/-/blob/internal/search/zoekt/indexed_search.go?L54-61) when partitioning repos. 

> Rather than blocking until all refs are resolved for all repos.

This is a valid concern, but not anything that should be effectively changed by this PR. As mentioned above, we already resolve all non-HEAD revisions during resolution, so the only thing this changes is we now expand globs to refs in this step too. Since globs are 1) less used, and 2) rarely used on more than one repo, and 3) only resolved if they're specified, this shouldn't change perf characteristics. 

Additionally, we now always resolve repos in pages, so this is only fully resolving a set of 500 repos before kicking off searcher jobs with the resolved set. I think we can make this even better by streaming resolved repos to jobs, but we're not there yet 😄

**Comment by keegancsmith** [comment]
cool, that all makes sense. This is what happens when I drive by review by mostly just reading the PR title and nothing else. My bad :)

**Review by rvantonder — APPROVED** [review]
> Again, sorry for the large change, but I was unable to figure out how to do this any more incrementally.

Someone once said, I forget who, "When one is doing a top-down rewrite one does not have the luxury to develop incrementally from scratch." Or something.

**Pull request merged**